lib/fs/walkfs.go

```diff
@@ -10,7 +10,55 @@
 
 package fs
 
-import "path/filepath"
+import (
+	"fmt"
+	"path/filepath"
+)
+
+func newAncestorDirList() *ancestorDirList {
+	l.Debugf("ancestorDirList: new")
+	return &ancestorDirList{list: make([]FileInfo, 0, 20)}
+}
+
+type ancestorDirList struct {
+	list []FileInfo
+}
+
+func (ancestors *ancestorDirList) AppendUnlessPresent(info FileInfo) bool {
+	l.Debugf("ancestorDirList: AppendUnlessPresent '%s'", info.Name())
+	for _, ancestor := range ancestors.list {
+		if SameFile(info, ancestor) {
+			return false
+		}
+	}
+	aLen := len(ancestors.list)
+	if aLen == cap(ancestors.list) {
+		t := make([]FileInfo, aLen, aLen*2)
+		copy(t, ancestors.list)
+		ancestors.list = t
+		l.Debugf("ancestorDirList: capacity increased to %d", cap(ancestors.list))
+	}
```

Comment:
If I read the specification of append correctly, it allocates only the space necessary for storing the result. Hence the allocated sizes would grow by 1 - 1, 2, 3, 4,... Not that it would hurt in this case. But I preferred the grows to be exponential - 1, 2, 4, 8,... I can delete this section of code if you like.

Reply:
Append already grows in a geometric fashion to get amortized O(1). You don't need to fight it.

Reply:
Fine, I probably misread the docs. I'll remove the excess code.

Reply:
fixed